sdk/iot/provisioning/src/az_iot_provisioning_client.c

```diff
@@ -47,161 +54,420 @@ AZ_NODISCARD az_result az_iot_provisioning_client_init(
+  AZ_PRECONDITION_NOT_NULL(mqtt_topic_filter);
+  AZ_PRECONDITION(mqtt_topic_filter_size > 0);
 
-  int32_t required_length
-      = az_span_size(str_dps) + az_span_size(str_registrations) + az_span_size(str_res) + 1;
+  az_span mqtt_topic_filter_span
+      = az_span_init((uint8_t*)mqtt_topic_filter, (int32_t)mqtt_topic_filter_size);
+  int32_t required_length = az_span_size(str_dps_registrations_res) + 1;
 
-  AZ_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_topic_filter, required_length);
+  AZ_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_topic_filter_span, required_length);
 
-  az_span remainder = az_span_copy(mqtt_topic_filter, str_dps);
-  remainder = az_span_copy(remainder, str_registrations);
-  remainder = az_span_copy(remainder, str_res);
-  az_span_copy_u8(remainder, '#');
+  az_span remainder = az_span_copy(mqtt_topic_filter_span, str_dps_registrations_res);
+  remainder = az_span_copy_u8(remainder, '#');
+  remainder = az_span_copy_u8(remainder, '\0');
 
-  *out_mqtt_topic_filter = az_span_slice(mqtt_topic_filter, 0, required_length);
+  if (out_mqtt_topic_filter_length)
+  {
+    *out_mqtt_topic_filter_length = (size_t)required_length;
+  }
 
   return AZ_OK;
 }
 
-// topic: $dps/registrations/res/202/?$rid=1&retry-after=3
-// payload:
-// {"operationId":"4.d0a671905ea5b2c8.42d78160-4c78-479e-8be7-61d5e55dac0d","status":"assigning"}
-AZ_NODISCARD az_result az_iot_provisioning_client_received_topic_payload_parse(
+AZ_INLINE az_iot_provisioning_client_registration_state registration_state_default()
+{
+  return (az_iot_provisioning_client_registration_state){ .assigned_hub_hostname = AZ_SPAN_NULL,
+                                                          .device_id = AZ_SPAN_NULL,
+                                                          .status = AZ_IOT_STATUS_OK,
+                                                          .extended_error_code = 0,
+                                                          .error_message = AZ_SPAN_NULL,
+                                                          .error_tracking_id = AZ_SPAN_NULL,
+                                                          .error_timestamp = AZ_SPAN_NULL };
+}
+
+/*
+Documented at
+https://docs.microsoft.com/en-us/rest/api/iot-dps/runtimeregistration/registerdevice#deviceregistrationresult
+  "registrationState":{
+    "x509":{},
+    "registrationId":"paho-sample-device1",
+    "createdDateTimeUtc":"2020-04-10T03:11:13.0276997Z",
+    "assignedHub":"contoso.azure-devices.net",
+    "deviceId":"paho-sample-device1",
+    "status":"assigned",
+    "substatus":"initialAssignment",
+    "lastUpdatedDateTimeUtc":"2020-04-10T03:11:13.2096201Z",
+    "etag":"IjYxMDA4ZDQ2LTAwMDAtMDEwMC0wMDAwLTVlOGZlM2QxMDAwMCI="}}
+*/
+AZ_INLINE az_result az_iot_provisioning_client_payload_error_code_parse(
+    az_json_token_member* tm,
+    az_iot_provisioning_client_registration_state* out_state)
+{
+  if (az_span_is_content_equal(AZ_SPAN_FROM_STR("errorCode"), tm->name))
```

Comment:
This is an incomplete comparison for JSON strings. Doing a byte-by-byte comparison for a known string won't catch escaped strings, where they are also semantically equal.

Either `tm->name` should be the unescaped string value, OR the json APIs should provide a comparison helper for callers.

Ideally: `"errorCode" == "\u0065rrorCode"`

For example, see the remark of the .NET API: https://docs.microsoft.com/en-us/dotnet/api/system.text.json.utf8jsonreader.valuetextequals?view=netcore-3.1#System_Text_Json_Utf8JsonReader_ValueTextEquals_System_ReadOnlySpan_System_Byte__

This is something we need to fix in the JSON API - we should file an issue. cc @JeffreyRichter, @vhvb1989

Reply:
Yes, it would be nice to fix this (if easy/simple)

Reply:
https://github.com/Azure/azure-sdk-for-c/issues/608

Reply:
We'd like an option to disable that functionality and use no-copy / no transform operations (this would be one of those cases).
We would be relying on IoT protocol specification and save some CPU cycles for small, battery-operated MCUs.

Reply:
is it guaranteed that the service will never return a JSON payload that has escaped strings in them? Let's follow the data end-to-end to make sure. I am surprised you'd want to turn this off (in which case, we might not want to implement it at all). Is that safe to do? We can try to make the overhead as low as possible, in the case where unescaping isn't needed, but I want to make sure we do the correct thing first and capture requirements before discussing perf.

Reply:
> is it guaranteed that the service will never return a JSON payload that has escaped strings in them?

I didn't say that. What I said was that all Azure IoT service _protocols_ do not have any escaped characters within them.
This means that the feature will be disabled for our protocol-generated json portions. I.e. our parsers would skip the feature as it's never needed.

That said, users can add json objects (sub-sections) to the larger document. For example the "payload" element - https://github.com/Azure/azure-sdk-for-c/pull/590/files#diff-5392e071b1599b9fab2ffa611e15d224R102.

In these cases, customers may choose to enable the feature to allow escaped chars as well as any UTF-8 encoded multi-byte chars.